innerprod/fullysec/paillier.go

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright (c) 2018 XLAB d.o.o
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fullysec
+
+import (
+	"fmt"
+	"math/big"
+
+	emmy "github.com/xlab-si/emmy/crypto/common"
+
+	"crypto/rand"
+
+	"github.com/fentec-project/gofe/data"
+	"github.com/fentec-project/gofe/internal"
+	"github.com/fentec-project/gofe/sample"
+)
+
+// paillerParams represents parameters for the fully secure Paillier scheme.
+type paillerParams struct {
+	l       int        // Length of data vectors for inner product
+	n       *big.Int   // a big integer, a product of two safe primes
+	nSquare *big.Int   // n^2 a modulus for computations
+	boundX  *big.Int   // a bound on the entries of the input vector
+	boundY  *big.Int   // a bound on the entries of the inner product vector
+	sigma   *big.Float // the standard deviation for the sampling a secret key
+	lambda  int        // safety parameter
+	g       *big.Int   // generator of the 2n-th residues subgroup of Z_n^2*
+}
+
+// Paillier represents a scheme based on the Paillier encryption.
+type Paillier struct {
+	Params *paillerParams
+}
+
+// NewPaillier configures a new instance of the scheme.
+// It accepts the length of input vectors l, security parameter lambda,
+// the bit length of prime numbers (giving security to the scheme, it
+// should be such that factoring two primes with such a bit length takes
+// at least 2^lambda operations), and boundX and boundY by which
+// coordinates of input vectors and inner product vectors are bounded.
+//
+// It returns an error in the case the scheme could not be properly
+// configured, or if the precondition boundX, boundY < (n / l)^(1/2)
+// is not satisfied.
+func NewPaillier(l, lambda int, bitLen int, boundX, boundY *big.Int) (*Paillier, error) {
```

Comment:
You can omit the `int` declaration after `lambda` since `bitLen` is also of that type.